pkg/controllers/externaloidc/externaloidc_controller.go

```diff
@@ -0,0 +1,363 @@
+	}
+
+	if existingCM != nil && existingCM.Data[authConfigDataKey] == string(authConfigJSON) {
+		return nil
+	}
+
+	errList := validateAuthenticationConfiguration(*authConfig)
+	if len(errList) > 0 {
+		return fmt.Errorf("auth config validation failed: %v", errList)
+	}
+
+	cm := corev1ac.ConfigMap(targetAuthConfigCMName, targetConfigMapNamespace).WithData(map[string]string{authConfigDataKey: string(authConfigJSON)})
+	if _, err := c.configMaps.ConfigMaps(targetConfigMapNamespace).Apply(ctx, cm, metav1.ApplyOptions{FieldManager: c.name}); err != nil {
+		return fmt.Errorf("could not apply changes to auth configmap %s/%s: %v", targetConfigMapNamespace, targetAuthConfigCMName, err)
+	}
+
+	syncCtx.Recorder().Eventf(c.eventName, "Synced auth configmap %s/%s", targetConfigMapNamespace, targetAuthConfigCMName)
+
+	return nil
+}
+
+// generateAuthConfig creates a structured JWT AuthenticationConfiguration for OIDC
+// from the configuration found in the authentication/cluster resource
+func (c *externalOIDCController) generateAuthConfig(auth configv1.Authentication) (*apiserverv1beta1.AuthenticationConfiguration, error) {
+	authConfig := apiserverv1beta1.AuthenticationConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AuthenticationConfiguration",
+			APIVersion: "apiserver.config.k8s.io/v1beta1",
+		},
+	}
+
+	for _, provider := range auth.Spec.OIDCProviders {
+		jwt := apiserverv1beta1.JWTAuthenticator{
+			Issuer: apiserverv1beta1.Issuer{
+				URL:                 provider.Issuer.URL,
+				AudienceMatchPolicy: apiserverv1beta1.AudienceMatchPolicyMatchAny,
+			},
+			ClaimMappings: apiserverv1beta1.ClaimMappings{
+				Username: apiserverv1beta1.PrefixedClaimOrExpression{
+					Claim: provider.ClaimMappings.Username.Claim,
+				},
+				Groups: apiserverv1beta1.PrefixedClaimOrExpression{
+					Claim:  provider.ClaimMappings.Groups.Claim,
+					Prefix: &provider.ClaimMappings.Groups.Prefix,
+				},
+			},
+		}
+
+		if len(provider.Issuer.Audiences) > 0 {
+			jwt.Issuer.Audiences = make([]string, 0, len(provider.Issuer.Audiences))
+			for _, aud := range provider.Issuer.Audiences {
+				jwt.Issuer.Audiences = append(jwt.Issuer.Audiences, string(aud))
+			}
+		}
+
+		if len(provider.Issuer.CertificateAuthority.Name) > 0 {
+			caConfigMap, err := c.configMapLister.ConfigMaps(caBundleConfigMapNamespace).Get(provider.Issuer.CertificateAuthority.Name)
+			if err != nil {
+				return nil, fmt.Errorf("could not retrieve auth configmap %s/%s to check CA bundle: %v", caBundleConfigMapNamespace, targetAuthConfigCMName, err)
+			}
+
+			caData, ok := caConfigMap.Data["ca-bundle.crt"]
+			if !ok || len(caData) == 0 {
+				return nil, fmt.Errorf("configmap %s/%s key \"ca-bundle.crt\" missing or empty", caBundleConfigMapNamespace, provider.Issuer.CertificateAuthority.Name)
+			}
+
+			jwt.Issuer.CertificateAuthority = caData
+		}
+
+		switch provider.ClaimMappings.Username.PrefixPolicy {
```

Comment:
The `UsernamePrefixPolicy` type won't allow any other value: https://github.com/liouk/cluster-authentication-operator/blob/de4ac96b4cd7c9f8de3a966d69ab31035403decb/vendor/github.com/openshift/api/config/v1/types_authentication.go#L410

If you prefer, for completeness, I can add a default with an error.

Reply:
Ok, so there will be a validation error, when someone tries to create one with another value?

Reply:
Exactly, this will be rejected by the API when someone will try to edit the authentication/cluster resource.

Reply:
What if a new valid option is added in the future? Can there be a period of time during an upgrade from N to N+1 where there is a cluster-authentication-operator at N and a CRD at N+1? I would include a default case to avoid all doubt.